File: app/src/main/java/com/stepyen/xlearn/fragment/basics/network/OkhttpFragment.java
package com.stepyen.xlearn.fragment.basics.network;

import android.util.Log;
import android.view.View;

import com.stepyen.xlearn.R;
import com.stepyen.xlearn.base.BaseTestFragment;
import com.xuexiang.xpage.annotation.Page;
import java.io.IOException;
import butterknife.OnClick;
import okhttp3.Call;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/**
 * date：2020-01-13
 * author：stepyen
 * description：
 */

@Page(name = "Okhttp")
public class OkhttpFragment extends BaseTestFragment {
    private static final String TAG = "OkhttpFragment";
    @Override
    public void initLayoutView() {
        addView(R.layout.fragment_okhttp);
    }

    @Override
    protected void initViews() {

    }


    @OnClick({R.id.btn_post})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.btn_post:
                postImage();

                break;
        }
    }


    public void postImage() {
        String url = "https://betachinese.kid58.com/AllConfig/allimage?appid=a3db11e896030050&md5=6e9b2c9f7866f32e0519a67f7ad31dd9";
        String clientInfo = "QauYefZ/GDHc5XZAt8M9qGBIiRa62GBXBxgpfYk4JU6hCcfoGmqQ1bGSlilmtDD7KIiIlN2XrEXZqdyP/QwsTX9q/+c44ATCKQqwBQJxK5OKsTHxHcw0e4my0igTFASTCi5l5SDwYUKHQ+Z9WqPAtJSYaz+hfNXrf6NNS3lmPeubrnmG4K89GU6AoOQizAHNonzlMcQjp6CT/dYyd62oTHPaUIov9rriR0lXQToVf6JJzYcxf/AeThYhGBQE4MBLeenpkXVcRpyHbdLH7iJ5qcb38WE08yIZT1U+QYiBOrSMZGbh2ULHccGKsInlviF3C+IKsxBNUm7Lzx1mT0i691a8gTJpB9Tb3MDY+IgP3epvTQqn19KOa85u/8yZGmsjZwey6voS9tnSK1BnpW8/hvpg5AIssqJp3FH0oTObD8o9gYPQtGhosGioZ8x29DyJqSF5JvjRPQB6RCgpGAKfWGjcM7vizJkpNe/fpKyYZvclpBYczKZ/+/6kDRxr1ydVBIgVlEHwL6VAVbdtnqHt9tm9hsf1Tp32qjMRFg==";

        OkHttpClient okHttpClient = new OkHttpClient();

        FormBody.Builder builder = new FormBody.Builder();
        FormBody formBody = builder.build();

        Request request = new Request.Builder()
                .url(url)
                .addHeader("client-info",clientInfo)
                .post(formBody)
                .build();

        okHttpClient.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                Log.d(TAG, "onFailure: "+e.toString());
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                Log.d(TAG, "onResponse: "+response.body().string());
            }
        });
    }

    public void post() {
        String url = "https://wanandroid.com/wxarticle/chapters/json";

        OkHttpClient okHttpClient = new OkHttpClient();

        FormBody.Builder builder = new FormBody.Builder();

        FormBody formBody = builder.build();

        Request request = new Request.Builder()
                .url(url)
                .post(formBody)
                .build();

        okHttpClient.newCall(request).enqueue(new okhttp3.Callback() {
            @Override
            public void onFailure(Call call, IOException e) {
                Log.d(TAG, "onFailure: "+e.toString());
            }

            @Override
            public void onResponse(Call call, Response response) throws IOException {
                Log.d(TAG, "onResponse: "+response.body());
                Log.d(TAG, "onResponse: "+response.body().toString());
                Log.d(TAG, "onResponse: "+response.body().string());
            }
        });

//        OkHttpUtils.post()
//                .url(url)
//                .build()
//                .execute(new StringCallback() {
//                    @Override
//                    public void onError(Call call, Exception e, int id) {
//                        Log.d(TAG, "onError: "+e.toString());
//                    }
//
//                    @Override
//                    public void onResponse(String response, int id) {
//                        Log.d(TAG, "onResponse: "+response);
//                    }
//                });
//
//
//        OkHttpUtils.post()
//                .url(url)
//                .build()
//                .execute(new Callback() {
//                    @Override
//                    public Object parseNetworkResponse(Response response, int id) throws Exception {
//                        Log.d(TAG, "parseNetworkResponse: 1 "+response.body().toString());
//                        Log.d(TAG, "parseNetworkResponse: 2 "+response.body().string());
//
//                        return response.body().string();
//                    }
//
//                    @Override
//                    public void onError(Call call, Exception e, int id) {
//
//                    }
//
//                    @Override
//                    public void onResponse(Object response, int id) {
//
//                    }
//                });

    }



}

File: app/src/main/java/com/stepyen/xlearn/fragment/basics/network/OkhttpFragment.kt
package com.stepyen.xlearn.fragment.basics.network

import android.graphics.BitmapFactory
import android.os.Handler
import com.orhanobut.logger.Logger
import com.stepyen.xlearn.App
import com.stepyen.xlearn.R
import com.stepyen.xlearn.base.BaseFragment
import com.stepyen.xutil.tip.ToastUtils
import com.xuexiang.xpage.annotation.Page
import kotlinx.android.synthetic.main.fragment_okhttp.*
import okhttp3.*
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit


/**
 * date：2020-01-13
 * author：stepyen
 * description：Okhttp 网络请求
 */
@Page(name = "Okhttp")
class OkhttpFragment : BaseFragment() {
    companion object {
        const val TAG = "OkhttpFragment"

        const val WHAT_DOWN_LOAD_IMAGE = 1


        const val IMAGE_URL_1 = "http://pic33.nipic.com/20131007/13639685_123501617185_2.jpg"

        const val IMAGE_URL_2 = "http://beta.admin.kid58.com/uploads/images/img/b5931961f98c985ed7bc4fc78c612a8b.png"

    }

    private var imageUrl = IMAGE_URL_1



    private val okHttpClient: OkHttpClient by lazy {

        OkHttpClient.Builder().apply {
            connectTimeout(20000L, TimeUnit.SECONDS)
            readTimeout(20000L, TimeUnit.SECONDS)
            writeTimeout(20000L, TimeUnit.SECONDS)
        }.build()


    }

    private val handle: Handler = Handler {

        when (it.what) {
            WHAT_DOWN_LOAD_IMAGE -> {
                val obj = it.obj as ByteArray
                val bitmap = BitmapFactory.decodeByteArray(obj, 0, obj.size)
                showIv.setImageBitmap(bitmap)
            }
            else -> {
            }

        }

        true
    }

    override fun initViews() {

    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_okhttp
    }


    override fun kotlinInitViews() {


        changeImagePath.setOnClickListener {
            imageUrl = if (imageUrl == IMAGE_URL_1) IMAGE_URL_2 else IMAGE_URL_1
        }

        getBtn.setOnClickListener {

        }
        postBtn.setOnClickListener {

        }
        postFormBtn.setOnClickListener {

        }
        postJsonBtn.setOnClickListener {

        }
        postDownLoadImageBtn.setOnClickListener {
            postDownLoadImage()
        }
        postDownLoadImageAndShowBtn.setOnClickListener {
            postDownLoadImageAndShow()
        }
        postUploadFileBtn.setOnClickListener {

        }

    }

    /**
     * 下载文件
     */


    /**
     * 下载图片并存在存储中
     */
    private fun postDownLoadImage() {

        val request = Request.Builder().apply { url(imageUrl) }.build()

        okHttpClient.newCall(request).enqueue(object : Callback {

            override fun onFailure(call: Call?, e: IOException?) {

                Logger.e(e.toString())

            }

            override fun onResponse(call: Call?, response: Response) {
                Logger.e(response.toString())

                val bytes = response.body()?.bytes()

                val imagePath = "test_image.jgp"

                FileUtil.writeToFile(App.get().externalCacheDir.absolutePath,imagePath,bytes)

                FileUtil.refreshSystemImageDataBase(context, File("${App.get().externalCacheDir}/${imagePath}"))

                Logger.e("保存成功：图片路径：${App.get().externalCacheDir}/${imagePath}")

                ToastUtils.toast("保存成功")
            }

        })
    }

    /**
     * 下载图片并显示
     */
    private fun postDownLoadImageAndShow() {


        val request = Request.Builder().apply { url(imageUrl) }.build()

        okHttpClient.newCall(request).enqueue(object : Callback {

            override fun onFailure(call: Call?, e: IOException?) {

                Logger.e(e.toString())

            }

            override fun onResponse(call: Call?, response: Response) {
                Logger.e(response.toString())

                val bytes = response.body()?.bytes()
                val message = handle.obtainMessage().apply {
                    obj = bytes
                    what = WHAT_DOWN_LOAD_IMAGE
                }

                handle.sendMessage(message)

            }

        })


    }


}